client/src/main/java/io/pravega/client/stream/impl/SegmentSelector.java

```diff
@@ -126,14 +130,22 @@ public Segment getSegmentForEvent(String routingKey) {
     }
 
     @Synchronized
-    private List<PendingEvent> updateSegmentsUponSealed(StreamSegments newStreamSegments, Segment sealedSegment,
+    private List<PendingEvent> updateSegmentsUponSealed(StreamSegmentsWithPredecessors successors, Segment sealedSegment,
                                                         Consumer<Segment> segmentSealedCallback) {
-        currentSegments = newStreamSegments;
-        createMissingWriters(segmentSealedCallback, newStreamSegments.getDelegationToken());
-        log.debug("Fetch unacked events for segment: {}, and adding new segments {}", sealedSegment, newStreamSegments);
-        List<PendingEvent> toResend = writers.get(sealedSegment).getUnackedEventsOnSeal();
-        writers.remove(sealedSegment); //remove this sealed segment writer.
-        return toResend;
+        if (successors == null) {
+            //stream is deleted, hence successors is null.
+            writers.get(sealedSegment).getUnackedEventsOnSeal()
+                    .forEach(pendingEvent -> pendingEvent.getAckFuture()
+                            .completeExceptionally(new NoSuchSegmentException(sealedSegment.toString())));
+            return Collections.emptyList();
+        } else {
+            currentSegments = currentSegments.withReplacementRange(successors);
+            createMissingWriters(segmentSealedCallback, currentSegments.getDelegationToken());
+            log.debug("Fetch unacked events for segment: {}, and adding new segments {}", sealedSegment, currentSegments);
+            List<PendingEvent> toResend = writers.get(sealedSegment).getUnackedEventsOnSeal();
```

Comment:
Use `writers.remove` to both get and remove - so you only access the collection once.

Reply:
done.